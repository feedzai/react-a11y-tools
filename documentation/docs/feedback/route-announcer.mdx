---
sidebar_position: 2
title: Route Announcer
---

import BrowserOnly from "@docusaurus/BrowserOnly";
import { Subtitle, BrowserWindow, PropsTable } from "../../src/components/index";
import { DemoWithRouter } from "./demos";

<Subtitle>Announces page changes for assistive technologies.</Subtitle>

The `RouteAnnouncer` adds an ARIA live region that announces route changes.

This component stands at the top of the React tree, and it should preferably be the first or second rendered component below your Router. It actively listens for a change on the location's `pathname` (e.g. Navigating from the `/events` page to the `/contacts` page).

So, this component will wrap your content, and an ARIA live region HTML element will render alongside it. Whenever the location changes, it pushes a message onto that HTML element, thus triggering the screen-reader by making it read the content out loud (e.g.: `Navigated to Events`).

To avoid the user having to insert the message every time the location changes manually, it queries the DOM for common elements and patterns that might signal something that can identify the new page.

These are:

- `h1` headings (there should be only one per page);
- Text content inside the `title` tag in the `head` element;
- Or, as a fallback, the pathname itself.

<BrowserWindow>
	<BrowserOnly>{() => <DemoWithRouter />}</BrowserOnly>
</BrowserWindow>
<br />

:::note How to

- Start your screen-reader (like `VoiceOver` or `Narrator`)
- Try navigating to any page visible on the demo.
- The screen reader will announce any page changes, like "Navigated to The Best Excuses To Spend A Cozy Valentineâ€™s Day In".
:::

### Usage

To take advantage of this component, place it as above as you can inside your React tree so that it can wrap your entire content. Since it will need to watch your router's location updates, try to put it right underneath your Router component.

```jsx
<Router>
	<RouteAnnouncer pathname={location.pathname}>
		<Home path="/" />
		<Company path="/company" />
		<Contacts path="/contacts" />
		<Shop path="/product" />
	</RouteAnnouncer>
</Router>
```

There are Codesandbox examples below for the most common use cases, like with Next.js own routing system, or with the super-popular React Router.

#### Using Next.js

Zeit's Next.js uses its routing mechanism, which users might find that as one of its greatest strengths. However, there are precise ways to handle tasks, such as adding functionality available on all pages by being an opinionated framework. Our component is one of them.

##### 1. Edit the `_app.js` file

To add the `RouteAnnouncer` component on Next, let's first go to the `pages/_app.js` file.
At the top, import the `useRouter` hook, which accesses the router object inside your app's function component. Since you're at it, also import our component.

```jsx
import { useRouter } from "next/router";
import { RouteAnnouncer } from "@jtmdias/react-a11y-tools/";
```

##### 2. Get the `pathname` from the `useRouter` hook

Then, inside the App function, reference the hook and destructure the object to obtain the `pathname` property. We need it to make the component listen for any significant location changes.

```jsx
export default function App({ Component, pageProps }) {
	const { pathname } = useRouter();
```

##### 3. Pass the `pathame` to the `RouteAnnouncer`

Finally, in the return statement, wrap the `Component` with the `RouteAnnouncer` and pass the `pathname` as a prop.

```jsx
return (
	<RouteAnnouncer pathname={pathname}>
		<Component {...pageProps} />
	</RouteAnnouncer>
);
```

Since users can frequently customize this file to override and control the page initialization, make sure that you place the component as high as you can.

That's it!

ðŸ”— Checkout the example [on this Codesandbox](https://codesandbox.io/s/react-a11y-tools-nextjs-0jxkg?file=/pages/_app.js)

#### Using `@reach/router`

`React Router`'s accessible brother, `Reach Router` is already built with accessibility in mind.

##### 1. Create a basic wrapper

First, let's create a component wrapper to listen for the router changes.
This will sit on top of the rest of the content (or wherever possible).

```jsx
import React from "react";
import { RouteAnnouncer } from "@jtmdias/react-a11y-tools/";
import { createHistory, LocationProvider } from "@reach/router";

export const history = createHistory(window);

export default function PageAnnouncer({ children }) {
	return (
		<LocationProvider history={history}>
			{({ location }) => {
				return <RouteAnnouncer pathname={location.pathname}>{children}</RouteAnnouncer>;
			}}
		</LocationProvider>
	);
}
```

This component basically listens for any location changes and updates the `RouteAnnouncer` with a new pathname.

##### 2. Find your main component

Second, let's find the file on your project that is the initial component or the one highest on the tree.
You know, something like `App.js` or `index.js`.
The one you pass into:

```jsx
import App from "../components/App";

const rootElement = document.getElementById("root");
ReactDOM.render(
	<StrictMode>
		<App />
	</StrictMode>,
	rootElement,
);
```

##### 3. Wrap your content

Got it? Great!
Once there, wrap everything with our newly created component.

```jsx
export default function App() {
	return (
		<PageAnnouncer>
			<div className="your-app">
				<ThisIsAHeader />
				<YourRouterComponentIsHere />
			</div>
		</PageAnnouncer>
	);
}
```

That's it?
Yep, that's everything you need to do.

ðŸ”— Checkout the live example [on this Codesandbox](https://codesandbox.io/s/react-a11y-tools-reachrouter-d7csq?file=/components/App.js)

#### Using `react-router`

The process of using the `RouteAnnouncer` with React Router is very similar to the one with `@reach/router`.
Nonetheless, here's what you need to do.

##### 1. Create a basic wrapper

First, let's create a component wrapper to listen for the router changes.
This will sit on top of the rest of the content (or wherever possible).

```jsx
import React from "react";
import { RouteAnnouncer } from "@jtmdias/react-a11y-tools/";
import { useLocation } from "react-router-dom";

export default function PageAnnouncer({ children }) {
	const { pathname } = useLocation();
	return <RouteAnnouncer pathname={pathname}>{children}</RouteAnnouncer>;
}
```

##### 2. Find your main component

Second, let's find the file on your project that is the initial component or the one highest on the tree.
You know, something like `App.js` or `index.js`.
The one you pass into:

```jsx
import App from "../components/App";

const rootElement = document.getElementById("root");
ReactDOM.render(
	<StrictMode>
		<App />
	</StrictMode>,
	rootElement,
);
```

##### 3. Wrap your content

Got it? Great!
Once there, wrap everything below the `Router` with our newly created component.

```jsx
import React from "react";
import { Header } from "./header";
import { BrowserRouter as Router } from "react-router-dom";
import AllThePageRoutesHere from "./page-router";
import PageAnnouncer from "./page-announcer";

export default function App() {
	return (
		<Router>
			<PageAnnouncer>
				<div className="your-app">
					<Header />
					<AllThePageRoutesHere />
				</div>
			</PageAnnouncer>
		</Router>
	);
}
```

This one was even simpler than `@reach/router`!

ðŸ”— Checkout the live example [on this Codesandbox](https://codesandbox.io/s/react-a11y-tools-react-router-03196?file=/components/App.js:0-529)

### Props

<PropsTable name="RouteAnnouncer" />
