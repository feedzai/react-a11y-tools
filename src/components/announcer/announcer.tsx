/**
 * This file is open-source. This means that it can be reproduced in whole
 * or in part, stored in a retrieval system transmitted in any form, or by
 * any means electronic with my prior permission as an author and owner
 * Please refer to the terms of the license agreement in the root of the project
 *
 * (c) 2021 joaodias.me
 */
import React, { NamedExoticComponent, memo } from "react";

export interface IAnnouncerProps {
	id?: string;
	ariaLive?: React.AriaAttributes["aria-live"];
	ariaAtomic?: React.AriaAttributes["aria-atomic"];
	styles?: React.CSSProperties;
	text?: string;
}

const styles: React.CSSProperties = {
	position: `absolute`,
	top: 0,
	width: 1,
	height: 1,
	padding: 0,
	overflow: `hidden`,
	clip: `rect(0, 0, 0, 0)`,
	whiteSpace: `nowrap`,
	border: 0,
};

const defaultProps: Partial<IAnnouncerProps> = {
	id: "rat-route-announcer",
	styles,
	ariaLive: "assertive",
	ariaAtomic: "true",
};

/**
 * Basic Announcer HTML element that tells a screen-reader of parts of the content that need the users attention.
 * Depending on the type of "politeness", users can be interrupted/or not by these messages.
 *
 * @example
 * // Changing page routes on a single-page application.
 * "Navigated to Create Account"
 *
 * // Status messages for form-related operations
 * "Your email is on the way!"
 * "Sorry, you need to fill the password field"
 *
 * // Notifications
 * "There's a new message"
 * "You're currently offline. Check your internet connection."
 *
 * @param {IAnnouncerProps} props
 * @returns {JSX.Element}
 */
export const Announcer: NamedExoticComponent<IAnnouncerProps> = memo(
	({
		id = defaultProps.id,
		styles = defaultProps.styles,
		ariaLive = defaultProps.ariaLive,
		ariaAtomic = defaultProps.ariaAtomic,
		text = defaultProps.text,
	}) => {
		/**
		 * Renders the contents inside the announcer div
		 *
		 * @returns {string | null}
		 */
		function renderText() {
			if (!text || text.length === 0) {
				return null;
			}

			return text;
		}

		return (
			<div
				id={id}
				aria-live={ariaLive}
				aria-atomic={ariaAtomic}
				className="rat__announcer"
				data-testid="rat-announcer"
				style={styles}
			>
				{renderText()}
			</div>
		);
	},
);
