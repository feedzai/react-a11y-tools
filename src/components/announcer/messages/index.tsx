/**
 * This file is open-source. This means that it can be reproduced in whole
 * or in part, stored in a retrieval system transmitted in any form, or by
 * any means electronic with my prior permission as an author and owner
 * Please refer to the terms of the license agreement in the root of the project
 *
 * (c) 2021 joaodias.me
 */
export interface ISetMessage {
	text: string;
	politeness?: "assertive" | "polite";
}
export interface IMessagesAnnouncerContext {
	message: string;
	politeness: "assertive" | "polite";
	setMessage: (message: ISetMessage) => void;
}

export type TUseMessagesAnnouncerReturns = (message: ISetMessage) => void;

export * from "./provider";
export * from "./useMessagesAnnouncer";
