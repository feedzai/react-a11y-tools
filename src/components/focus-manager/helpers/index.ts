/*
 * Please refer to the terms of the license
 * agreement.
 *
 * (c) 2024 Feedzai, Rights Reserved.
 */
import { FocusManagerProps } from "../types";
import { Options } from "focus-trap";

/**
 * Sets the initial focus element.
 */
export function getInitialFocus(
	autoFocus: FocusManagerProps["autoFocus"],
	initialFocus: Options["initialFocus"],
) {
	// If true, resumes the normal behaviour
	if (autoFocus === true) {
		return undefined;
	}

	// If false, prevents the autoFocus from happening
	else if (autoFocus === false) {
		return false;
	}

	// If is nil, uses the value defined in the `initialFocus` param,
	// which if also nil, resumes the normal behaviour
	return initialFocus;
}

/**
 * Sets the fallback focus element to a:
 * 1. Defined option as prop
 * 2. A fallback ref element
 */
export function getFallbackFocus<GenericElement extends HTMLElement>(
	element: GenericElement | null,
	fallbackFocus: Options["fallbackFocus"],
) {
	if (fallbackFocus) {
		return fallbackFocus;
	}

	return element ?? undefined;
}

/**
 * Returns the correct prop value for the return focus functionality.
 */
export function getReturnFocusProps(restoreFocus: FocusManagerProps["restoreFocus"]) {
	if (restoreFocus && restoreFocus === true) {
		return {
			returnFocusOnDeactivate: true,
		};
	}

	return {
		setReturnFocus: restoreFocus,
	};
}
