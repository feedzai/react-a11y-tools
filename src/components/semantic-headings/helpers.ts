/*
 * Please refer to the terms of the license agreement.
 *
 * (c) 2023 joaodias.me, Rights Reserved.
 */
import { inRange } from "@jtmdias/js-utilities";
import { FIRST_HEADING_LEVEL, LAST_HEADING_LEVEL } from "./constants";

/**
 * Checks if the heading level is within range and valid.
 * If not:
 * 1.1. Outputs a warning to the browser's console (during development)
 * 1.2. and returns a value between the range (i.e: if invalid h7, then returns 6)
 *
 * @param {number} level
 * @returns {number}
 */
export function getHeadingLevel(level: number): number {
	const isWithinRange = inRange(level, FIRST_HEADING_LEVEL, LAST_HEADING_LEVEL);

	switch (true) {
		case isWithinRange:
			return level;

		default:
			return Math.min(Math.max(FIRST_HEADING_LEVEL, level), LAST_HEADING_LEVEL);
	}
}
