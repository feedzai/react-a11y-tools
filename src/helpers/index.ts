/* istanbul ignore file */
/**
 * This file is open-source. This means that it can be reproduced in whole
 * or in part, stored in a retrieval system transmitted in any form, or by
 * any means electronic with my prior permission as an author and owner
 * Please refer to the terms of the license agreement in the root of the project
 *
 * (c) 2021 joaodias.me
 */
export * from "./isBrowser";
export * from "./run-after-transition";
export { focusWithoutScrolling } from "./focus-without-scrolling";
