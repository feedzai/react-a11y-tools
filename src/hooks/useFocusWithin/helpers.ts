/* istanbul ignore file */
/*
 * The file cannot be reproduced in whole or in part, stored in a retrieval system, transmitted
 * in any form, or by any means electronic, mechanical, or otherwise, without
 * the prior permission of the owner. Please refer to the terms of the license
 * agreement.
 *
 * (c) 2023 joaodias.me, Rights Reserved.
 */
import { FocusEvent as ReactFocusEvent, useCallback, useRef } from "react";
import { useSafeLayoutEffect } from "../index";

export class SyntheticFocusEvent implements ReactFocusEvent {
	nativeEvent: FocusEvent;
	target: Element;
	currentTarget: Element;
	relatedTarget: Element;
	bubbles: boolean;
	cancelable: boolean;
	defaultPrevented: boolean;
	eventPhase: number;
	isTrusted: boolean;
	timeStamp: number;
	type: string;

	constructor (type: string, nativeEvent: FocusEvent) {
		this.nativeEvent = nativeEvent;
		this.target = nativeEvent.target as Element;
		this.currentTarget = nativeEvent.currentTarget as Element;
		this.relatedTarget = nativeEvent.relatedTarget as Element;
		this.bubbles = nativeEvent.bubbles;
		this.cancelable = nativeEvent.cancelable;
		this.defaultPrevented = nativeEvent.defaultPrevented;
		this.eventPhase = nativeEvent.eventPhase;
		this.isTrusted = nativeEvent.isTrusted;
		this.timeStamp = nativeEvent.timeStamp;
		this.type = type;
	}

	isDefaultPrevented(): boolean {
		return this.nativeEvent.defaultPrevented;
	}

	preventDefault(): void {
		this.defaultPrevented = true;
		this.nativeEvent.preventDefault();
	}

	stopPropagation(): void {
		this.nativeEvent.stopPropagation();
		this.isPropagationStopped = () => true;
	}

	isPropagationStopped(): boolean {
		return false;
	}

	// eslint-disable-next-line @typescript-eslint/no-empty-function
	persist() { }
}

export function useSyntheticBlurEvent(onBlur: (e: ReactFocusEvent) => void) {
	const stateRef = useRef({
		isFocused: false,
		onBlur,
		observer: null as MutationObserver | null,
	});
	stateRef.current.onBlur = onBlur;

	// Clean up MutationObserver on unmount. See below.
	// eslint-disable-next-line arrow-body-style
	useSafeLayoutEffect(() => {
		const state = stateRef.current;
		return () => {
			if (state.observer) {
				state.observer.disconnect();
				state.observer = null;
			}
		};
	}, []);

	// This function is called during a React onFocus event.
	return useCallback((event: ReactFocusEvent) => {
		// React does not fire onBlur when an element is disabled. https://github.com/facebook/react/issues/9142
		// Most browsers fire a native focusout event in this case, except for Firefox. In that case, we use a
		// MutationObserver to watch for the disabled attribute, and dispatch these events ourselves.
		// For browsers that do, focusout fires before the MutationObserver, so onBlur should not fire twice.
		if (
			event.target instanceof HTMLButtonElement ||
			event.target instanceof HTMLInputElement ||
			event.target instanceof HTMLTextAreaElement ||
			event.target instanceof HTMLSelectElement
		) {
			stateRef.current.isFocused = true;

			const target = event.target;
			const onBlurHandler = (event: FocusEvent) => {
				stateRef.current.isFocused = false;

				if (target.disabled) {
					// For backward compatibility, dispatch a (fake) React synthetic event.
					stateRef.current.onBlur?.(new SyntheticFocusEvent("blur", event));
				}

				// We no longer need the MutationObserver once the target is blurred.
				if (stateRef.current.observer) {
					stateRef.current.observer.disconnect();
					stateRef.current.observer = null;
				}
			};

			target.addEventListener("focusout", onBlurHandler as EventListenerOrEventListenerObject, { once: true });

			stateRef.current.observer = new MutationObserver(() => {
				if (stateRef.current.isFocused && target.disabled) {
					stateRef.current.observer?.disconnect();
					target.dispatchEvent(new FocusEvent("blur"));
					target.dispatchEvent(new FocusEvent("focusout", { bubbles: true }));
				}
			});

			stateRef.current.observer.observe(target, {
				attributes: true,
				attributeFilter: ["disabled"],
			});
		}
	}, []);
}
