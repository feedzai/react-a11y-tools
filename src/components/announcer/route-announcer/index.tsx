/*
 * Please refer to the terms of the license
 * agreement.
 *
 * (c) 2023 joaodias.me, Rights Reserved.
 */
import React, { FunctionComponent, useCallback, useEffect, useState } from "react";
import { usePrevious } from "@jtmdias/js-utilities";
import { Announcer } from "../announcer";

export interface IRouteAnnouncerActions {
	/**
	 * The action predicate.
	 * @default "Navigated To"
	 */
	navigation: string;

	/**
	 * The fallback predicate for actions.
	 */
	fallback: string;
}
export interface IRouteAnnouncerProps {
	/**
	 * An optional id for the `focus wrapper` HTML element.
	 */
	id?: string;

	/**
	 * History-based Location pathnme
	 */
	pathname?: string;

	/**
	 * Content to be read by the screen-reader on the `aria-live` announcer
	 */
	action?: IRouteAnnouncerActions;
}

const DEFAULT_WRAPPER_ID = "content-focus-wrapper";

export const defaultProps: Partial<IRouteAnnouncerProps> = {
	id: DEFAULT_WRAPPER_ID,
	action: {
		navigation: "Navigated to",
		fallback: "new page at",
	},
};

/**
 * Queries the HTML Head to find the `title` tag
 *
 * @returns {boolean}
 */
export function hasDocumentTitle(): boolean {
	return !!document.title;
}

/**
 * Queries the DOM in order to find the first h1.
 *
 * @param {string} id
 * @returns {string | undefined}
 */
export function getHeadingText(id: string): string | undefined {
	const element: HTMLHeadingElement | null = document.querySelector(`#${id} h1`);

	return element && element.textContent ? element.textContent : undefined;
}

/**
 * The `RouteAnnouncer` is a wrapper for the app's content,
 * as well as the `Announcer` component.
 *
 * It listens for a change in the `pathname` prop, so that it passes a new text
 * for the `Announcer` to read.
 *
 * @example
 * // Passing a new pathname as prop
 * <RouteAnnouncer pathname="/create-account"></RouteAnnouncer>
 *
 * // The screen reader outputs:
 * "Navigated to Create Account"
 *
 * @param {IRouteAnnouncerProps} props
 * @returns {JSX.Element}
 */
export const RouteAnnouncer: FunctionComponent<IRouteAnnouncerProps> = ({
	id = defaultProps.id,
	pathname,
	action = defaultProps.action,
	children,
}) => {
	const [text, setText] = useState("");
	const previousPathname = usePrevious(pathname);

	/**
	 * Creates the string to be read by the `Announcer` component
	 * It depends on the existance of:
	 * 1. The location pathname.
	 * 2. A title in the document head.
	 * 3. The first h1 it finds on the DOM.
	 *
	 * Updates the state with whatever comes last.
	 *
	 * @returns {void}
	 */
	const setAnnouncerText = useCallback(() => {
		const hasTitle = hasDocumentTitle();
		/* istanbul ignore next */
		const firstHeading = getHeadingText(id || DEFAULT_WRAPPER_ID);

		let pageName = `${(action as IRouteAnnouncerActions).fallback} ${pathname}`;

		if (hasTitle) {
			pageName = document.title;
		}

		if (typeof firstHeading === "string" && firstHeading.length > 1) {
			pageName = firstHeading;
		}

		const newMessage = `${(action as IRouteAnnouncerActions).navigation} ${pageName}`;

		setText(newMessage);
	}, [action, id, pathname]);

	useEffect(() => {
		if (previousPathname && previousPathname !== pathname) {
			setAnnouncerText();
		}
	}, [pathname, previousPathname, setAnnouncerText]);

	return (
		<div
			id={id}
			style={{
				outline: "none",
			}}
			tabIndex={-1}
		>
			{children}
			<Announcer message={text} />
		</div>
	);
};
