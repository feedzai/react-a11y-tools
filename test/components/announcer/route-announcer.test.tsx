/**
 * This file is open-source. This means that it can be reproduced in whole
 * or in part, stored in a retrieval system transmitted in any form, or by
 * any means electronic with my prior permission as an author and owner
 * Please refer to the terms of the license agreement in the root of the project
 *
 * (c) 2021 joaodias.me
 */
import { Link, RouteComponentProps, Router, useLocation } from "@reach/router";
import "@testing-library/jest-dom";
import { render, screen, waitFor } from "@testing-library/react";
import React, { FunctionComponent, useEffect } from "react";
import {
	defaultProps,
	RouteAnnouncer,
	getHeadingText,
} from "../../../src/components/announcer/route-announcer";
import { IRouteAnnouncerProps } from "../../../src/components/announcer/route-announcer";
import { renderWithRouter } from "../../helpers/renderWithRouter";

const Home: FunctionComponent<RouteComponentProps> = () => {
	return (
		<div>
			<h1>Home</h1>
			<p>You are on the initial page</p>
		</div>
	);
};

const About: FunctionComponent<RouteComponentProps> = () => {
	return (
		<div>
			<h1>About</h1>
			<p>You are on the about page</p>
		</div>
	);
};

const Contacts: FunctionComponent<RouteComponentProps> = () => {
	return (
		<div>
			<p>You are on the contacts page</p>
		</div>
	);
};

const Work: FunctionComponent<RouteComponentProps> = () => {
	useEffect(() => {
		document.title = "Work";
	}, []);
	return (
		<div>
			<p>You are on the work page</p>
		</div>
	);
};
const NoMatch: FunctionComponent<RouteComponentProps> = () => {
	return (
		<div>
			<p>404: No Match</p>
		</div>
	);
};

const App = (): JSX.Element => {
	const location = useLocation();
	return (
		<RouteAnnouncer {...defaultProps} pathname={location.pathname}>
			<Link to="/">Home</Link>
			<Link to="/about">About</Link>
			<Link to="/work">Work</Link>
			<Link to="/contacts">Contacts</Link>
			<Router>
				<Home path="/" />
				<About path="/about" />
				<Contacts path="/contacts" />
				<Work path="/work" />
				<NoMatch default />
			</Router>
		</RouteAnnouncer>
	);
};

describe("<RouteAnnouncer />", () => {
	let props: IRouteAnnouncerProps;

	beforeEach(() => {
		props = {
			...defaultProps,
			pathname: "/",
		};
	});

	it("should render without errors", () => {
		const component = render(<RouteAnnouncer id="app-focus-wrapper" {...props} />);

		expect(component).toMatchSnapshot();
	});

	it("should find the heading on the document", () => {
		renderWithRouter(<App />);
		expect(getHeadingText("content-focus-wrapper")).toEqual("Home");
	});

	it("should not update the announcer on the first location", async () => {
		// 0. Render the app with the `router`
		renderWithRouter(<App />);

		// 2. Expect the new heading and paragraphs to exist
		expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent("Home");
		expect(screen.getByText("You are on the initial page")).toBeVisible();

		// 3. The Announcer should not have any text
		expect(screen.getByTestId("rat-announcer")).not.toHaveTextContent("Navigated to Home");
	});

	describe("should update the announcer when the location changes", () => {
		it("By it's heading", async () => {
			// 0. Render the app with the `router`
			const {
				history: { navigate },
			} = renderWithRouter(<App />);

			// 1. Transition to another page
			await navigate("/about");

			// 2. Expect the new heading and paragraphs to exist
			expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent("About");
			expect(screen.getByText("You are on the about page")).toBeVisible();

			// 3. The Announcer should have new text
			await waitFor(() => {
				expect(screen.getByText("Navigated to About")).toBeInTheDocument();
			});
		});

		it("By it's location", async () => {
			// 0. Render the app with the `router`
			const {
				history: { navigate },
			} = renderWithRouter(<App />);

			// 1. Transition to another page
			await navigate("/contacts");

			// 2. Expect the new heading and paragraphs to exist
			expect(screen.getByText("You are on the contacts page")).toBeVisible();

			// 3. The Announcer should have new text
			await waitFor(() => {
				expect(
					screen.getByText(
						`${defaultProps.action?.navigation} ${defaultProps.action?.fallback} /contacts`,
					),
				).toBeInTheDocument();
			});
		});

		it("By it's page title", async () => {
			// 0. Render the app with the `router`
			const {
				history: { navigate },
			} = renderWithRouter(<App />);

			// 1. Expect the heading and paragraph to exist
			expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent("Home");
			expect(screen.getByText("You are on the initial page")).toBeVisible();

			// 2. Transition to a new page
			await navigate("/work");

			// 3. Expect the new paragraph to exist
			expect(screen.getByText("You are on the work page")).toBeVisible();

			// 4. The Announcer should have new text set by the page's title.
			await waitFor(() => {
				expect(screen.getByText("Navigated to Work")).toBeInTheDocument();
			});
		});
	});
});
