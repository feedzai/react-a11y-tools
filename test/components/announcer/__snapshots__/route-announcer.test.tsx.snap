// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RouteAnnouncer /> should render without errors 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="content-focus-wrapper"
        style="outline: none;"
        tabindex="-1"
      >
        <div
          aria-atomic="true"
          aria-live="assertive"
          class="rat__announcer"
          data-testid="rat-announcer"
          id="rat-route-announcer"
          style="position: absolute; top: 0px; width: 1px; height: 1px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border: 0px;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="content-focus-wrapper"
      style="outline: none;"
      tabindex="-1"
    >
      <div
        aria-atomic="true"
        aria-live="assertive"
        class="rat__announcer"
        data-testid="rat-announcer"
        id="rat-route-announcer"
        style="position: absolute; top: 0px; width: 1px; height: 1px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border: 0px;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
